### DISCUSSION TITLE
Allow specifying inclusive sets of criteria when using the `relatedTo*` arguments in GraphQL

### DISCUSSION THREAD

**Discussion by andris-sevcenko** [craftcms/cms#9003] [Ideas]
### Description

As of #7528, it's possible to specify multiple sets of criteria which are then applied using "and" - namely, an element must be present in all of the criteria sets.

It would be great if it was possible to make this behave in an inclusive, rather an exclusive way, where an element has to be present in any of the criteria sets.

**Comment by joshuabaker** [comment]
Could this work as it does in Twig, where the first child is just a string (i.e. `["or", ...]`)? That works really nicely, including for nested criteria.

**Comment by andris-sevcenko** [comment]
That's not possible with GraphQL, unfortunately. It's possible to mix and match scalar values by some creative validation on the PHP end, but it's not possible to have a list of "scalar or complex" values.

The currently planned approach is to have a syntax like
``` graphql
{
    operation: "and", 
    criteria: [
        {entryCriteriaObject},
        {entryCriteriaObject}
    ]
}
```

**Reply by andris-sevcenko** [reply]
This is most likely pushed back to Craft 5.0. The reason for that is that Craft 5.0 will bring changes to the content model and how that is stored, which will in all likelihood (I'd estimate at least 90% chance) result in changes for GraphQL that will break almost all existing queries.

This change would too render all the existing queries that use the `relatedToEntries` (and the like) argument to break, as it's impossible for GraphQL to cast (simplifying) an `[Object]` input to a `{operation: String, criteria: [Object]}` signature.

So, in order to be as non-disruptive as possible, we're pushing this (and other planned changes) back to Craft 5.0, so that we don't have to ask developer to go over and adjust their GQL queries twice in a short timespan.

Meanwhile, if you really, really need this, you can add a new argument by [using the `registerGqlQueries`](https://github.com/craftcms/cms/blob/develop/src/services/Gql.php#L141) event to add additional arguments to existing queries and then adding [an argument handler to add some additional logic to those arguments](https://craftcms.stackexchange.com/questions/38012/how-do-i-write-a-graphql-query-to-search-through-sub-entries-entries-as-fields-a).

**Reply by joshuabaker** [reply]
Thanks for the follow up, @andris-sevcenko 🙏🏻
